### ISSUE TITLE
Bug: Pre-request script is executed before the request is prepared

### ISSUE THREAD

**Issue by pietrygamat** [usebruno/bruno#2249]
### I have checked the following:

- [X] I use the newest version of bruno.
- [X] I've searched existing issues and found nothing related to my issue.

### Describe the bug

When executing the request, _bruno_ goes through several stages of configuration, among them: execution o pre-request script, post-request script, and run of 'configureRequest` method. The current execution order is:
- **execute pre-request scripts**
- **configure request** (interpolate vars, sets proxy, auth headers. cookies)
- execute request
- execute post-request

I wonder if it would make more sense to execute the steps in this order:
- **configure request**
- **execute-pre-request** script
- execute request
- execute post-request script

This way any logic built into bruno may be overwritten by user's script and on top of that, user's script execute on request that is closer to what is actually being sent (e.g. with variables resolved, headers set).

For example, the below bru file logs this to console:
```
Before
{{api_url}}
{"Content-Type":"application/json"}

After
https://localhost:8083
{"Content-Type":"application/json","Authorization":"Basic dXNlcjpwYXNzd29yZA=="}
```
while I would expect the output Before and After to be the same.

### .bru file to reproduce the bug

meta {
  name: example
  type: http
}

post {
  url: {{api_url}}
  body: text
  auth: basic
}

headers {
  Content-Type: application/json
}

auth:basic {
  username: user
  password: password
}

script:pre-request {
  console.log('Before');
  console.log(req.url);
  console.log(JSON.stringify(req.getHeaders()));
}

script:post-response {
  console.log('After');
  console.log(req.url);
  console.log(JSON.stringify(req.getHeaders()));
}

### Screenshots/Live demo link

N/A

**Comment by centur** [comment]
Wow :( 
This bug broke heaps of my existing scripts, I've to find a older bruno version and get back to it, as I'm suddenly blocked .
@lohxt1 I wonder if these types of scripts still possible after the fix.

```
console.log(`createTenant`)
const { readFileSync } = require('fs');
const { URL } = require('url');
const path = require('path');

var attachmentFilename ="../../my-config.zip"
const attachmentPath = path.join(bru.cwd(), attachmentFilename);

console.log(attachmentPath);

const attachment = readFileSync(attachmentPath);
const attachmentLength = attachment.length;

req.setHeader("Content-Type", "application/zip");
req.setHeader("Content-Length", attachmentLength);
req.setBody(attachment);

var randomName = faker.location.city()
bru.setVar("configName", `bruno-${randomName}`);
```
For the context - because bruno does not support raw file (and I want some extra control over file selection) I've implemented this script to send raw binary content in post request. few versions ago it was fully functional script, now it's not working at all. Not sure what exactly was broken but errors or console logs from pre-req are not delivered anywhere and this file is not being sent to api eiter
And FYI  - console.log in pre-request ceased to work, nothing shows in console even if request is trivial

Bruno version: 1.17.0
MacOS: 14.4.1 (23E224)

**Comment by pietrygamat** [comment]
@lohxt1 do you have an update on this one, or a public branch where this can be peeked at?

> We have a fix in place for this, which is planned to be included in the upcoming release.

**Comment by centur** [comment]
@pietrygamat check if it's fixed in 1.18 which is just dropped in release channel. It's fixed my scripts with file uploads for sure.

**Comment by pietrygamat** [comment]
@lohxt1 
> Issue with configuring request before pre-request script : If the interpolation of the URL happens before the pre-request script, and if the URL is using a variable that is supposed to change in the pre-request script, the new updated variable value won't be replaced in the already interpolated URL.

This is a good point. I have been thinking under different assumptions: the script would be able to replace variables on those variables that have not been already replaced (no value provided in env file, etc), but I can see how this is not always the case. In that context proposed solution seems like a reasonable nice-to-have. 

That, however, does not solve my core issue here: the `req` accessed during pre-request script is not processed by `configure` method, which does a lot more than interpolation - most importantly, sets various fields related to authorization. That is something I hoped to rely on when implementing #2061 - my idea is that obtaining an access token should happen without overwriting user's request - during request configuration - as obtaining access token is functionally the same as for example setting basic auth header, which is also done in configure phase. 

So having a script execute on a fully prepared request would allow accessing anything that the general configuration sets for the user (including access_token) and would also theoretically prevent edge cases, where something set in the pre-request script is overwritten by default bruno behavior. But it is a sort of a chicken-egg problem, because one may claim they don't want to read anything set in configuration, but rather feed `configure` with required data... both approaches have their merits. 

If you vote to stay with current behavior, this may be closed.

**Comment by centur** [comment]
@lohxt1 yes it does, I tested it when I saw 1.18  was pushed to brew.sh 
https://github.com/usebruno/bruno/issues/2249#issuecomment-2126278862

Re `bru.interpolateVars();` - From the name it's not clear what would it do and what would happens if someone would call it twice (in a collection wide and request-specific pre-request section) - and then dev will be surprised with a result. 

With our samples above - it's clear that there are some scenarios when you want to change vars before interpolation (like inferring a full url or adjusting the path like in @pietrygamat  scenario) and some cases where we want to have a configured request and re-write parts of it (my file uploads workaround). Taking a step back from solution here I think end result should satisfy both behaviours:
 -  Pre-request script should be able to influence interpolation results (so it runs BEFORE final interpolation)
 -  Pre-request script should be able to change aspects of request before it's sent.

As an idea - sequence of steps can be something like this:

1. Request instance is initialized, but not yet interpolated or validated, it's available as `req.` in script. 
   - It can be  a shallow object which looks like Request, but I'd prefer OG request and take the risks of me messing it up for the flexibility it gives.
2. Pre-request script runs, can make changes to env vars and request data itself AND allows a way to omit final interpolation (`bru.skipInterpolation(req)` - calling this implies developer takes responsibility that they handled interpolation of all values and they don't want those to be changed ).
3. Request is finalized - all values are interpolated, it's validated, proxy set etc
4. Request sent.

So it's the sequence that is currently in place with an ability to force skip automatic interpolation, but not proxy\cookies settings. Thoughts ?
